### ISSUE TITLE
add exerciseSessions endpoint

### ISSUE THREAD

**Issue by majapetersen** [zeeguu-ecosystem/Zeeguu-API#65]
"exerciseSessions" object should include: 

- number of exercise sessions completed in the time period, 
- overall correctness of exercises completed in the time period, 
- time spent on exercises in the time period, 
- list of words practiced (mapping the word to its' translation and a list of typeOfExercise, correctness(first attempt, second attempt, etc., hints in percentage)),
- list of words translated but not included in exercises including their translations, 
- list of learned words (assuming all learned words are practiced words).

api.getExerciseSessions(studentID, cohortID, duration) - returns a list of "exerciseSessions" objects of what exercise sessions the student has completed within the time period.

**Comment by mircealungu** [comment]
re) re 1) ✅ number of actual exercises it is.

re) re) 2) 3) 4) ✅ I'll see what works best when I start implementing it

re re) 5) "words not be included in exercises" in a period can be not included for multiple reasons
- there was not enough time for them to be included, because the student only did an exercise for 1 minute say, but he has translated 40 words
- the system has deemed some of them "not fit for study" 
- the scheduling algo has not been given the chance to schedule these words

re) re) 6) I'm not sure I can explain what I mean without looking at the figma. And I didn't understand well your argument either :( So let me try again with an In extremist case to maybe make my case more clear. I've practiced three words in the last week. Each of them four times. Two I've learned. One I didn't. Then in the main view, the top view, where I see the exercise history for all the three words, the two words that I've learned will be highlighted in a green background or something similar. Everything else stays the same. So the teacher gets an understanding that besides having exercised them, I've also learned them. And we don't have to list these two words twice on the page: first with exercise history at the top, and second just as a simple list of learned words at the bottom. do you see what I mean?

**Comment by SaraTolstrupHusum** [comment]
Re our discussion at the meeting today and the solution that we agreed upon, we need 4 endpoints to show the students exercise activities to the teachers.
We will continue this issue for the general activity overview and create 3 separate for the others endpoints in a minute.

this endpoint should be: api.getExerciseActivityOverview(studentID, cohortID, duration) or whatever makes sense when you start implementing it and it should return the following:

- total time spent on exercises in the time period
- number of words practised in the time period
- % of exercises solved on first attempt in the time period
- number of learned words in the time period
- number of words that have never been exercised (excluded by algo or not yet scheduled) in the time period

**Comment by SaraTolstrupHusum** [comment]
@mircea
We still need this one.

**Comment by mircealungu** [comment]
thanks for the reminder! i'll try to handle it this afternoon.

**Comment by mircealungu** [comment]
Oki, turns out I still have a bit of time.

Question: it seems that in /student_activity_overview we already have points 1 and 3 above (total time, and % of first-attempt solutions). Wouldn't it make sense to simply add the extra info in there? 

![image](https://user-images.githubusercontent.com/464519/121493429-1fe2ac80-c9d8-11eb-888b-a0fc6cfa5182.png)

**Comment by SaraTolstrupHusum** [comment]
We are making the call to /student_activity_overview in a component completely unrelated to the one, where we need to make the api.getExerciseSessions call, so including the missing data in /student_activity_overview, won't really make much sense, as we will need to query the endpoint twice anyways and we will then end up getting data that we do not need on both calls.

Furthermore, after having implemented the api call to /student_words_not_studied yesterday, we realised that if a student has looked up an entire sentence all  bits of the sentence are added to the list as individual words.
eg: "quirido cambriar la letra para encajar ciertas cosas que estaban haciendo" figures on the list as:
"quirido cambriar" 
"quirido cambriar la letra" 
"quirido cambriar la letra para encajar" 
"quirido cambriar la letra para encajar ciertas" 
"quirido cambriar la letra para encajar ciertas cosas" 
"quirido cambriar la letra para encajar ciertas cosas que estaban haciendo"  (I can't remember the exact render, but you get the idea.)
As the teachers consistently underline their need for quick overview, we decided to filter the list and not show words/sentences that were only a subset of a longer version of the same sentence. 
It would be nice if we could the backend could do a similar filtering, so that the number of words/sentences not studied is similar to the number of words we end up rendering. (- It is only a problem with the "non-studied" word list, and you can see the filtering we are doing in NonStudiedWordsList on the newTeacher branch.)

**Comment by mircealungu** [comment]
**/student_activity_overview**

If an endpoint gives you more info, use the one you need, and don't worry about the extra info. 

What is also not nice is having two endpoints that overlap in their functionality. In the approach you propose we have some info about exercise overview in /student_activity and some of it is duplicated again in /student_exercise_activity.

Thus I only see two nice solutions: 
1. return everything in /student_activity_overview 
2. implement two disjoint endpints /student_reading_overview and /student_exercises_overview

The first one is nicer because you never have to call two endpoints to get all the data you need. And if you get more data, you just don't use it. 

** translation increments **
In the latest version of the reader on development the increments are not saved anymore. 

However, the DB is already populated with some of that data from the past that has the subsets and this is what you saw yesterday.

I guess for now it's ok that you're filtering the subsets. But we shouldn't implement any more special case handling on the backend. In the short term, the total number of words not studied will literally count the words not the phrases. And in the long term the problem will slowly disappear.

**Comment by SaraTolstrupHusum** [comment]
If the backend is already fixing the increment problem for the future. We will just remove our filtering (as there is a silly bug in it - it only works on every other render. #riddleMeThat?!?)
And if having just one endpoint and some excess data is the better solution... Then for sure, lets go with that.

**Comment by mircealungu** [comment]
Excellent! I didn't want to tell you to remove code, because you spent time working on it... but I'm always happy when we decrease the codebase size!

New endpoint is here: 
https://github.com/zeeguu-ecosystem/Zeeguu-API/blob/76b3917a91407e3a3a1aa96eac5eca9234c72aa0/zeeguu_api/api/teacher_dashboard/student_overviews.py#L12

**Comment by mircealungu** [comment]
I've done quite a big refactoring for this new version of the endpoint. I hope I didn't break anything. Tests are passing and I was as attentive as I could. But keep an eye open for endpoints that might be broken and let me know!

**Comment by SaraTolstrupHusum** [comment]
Will do.

**Comment by SaraTolstrupHusum** [comment]
There is a mismatch between the number of exercises returned by this endpoint and the number of exercises that we get in the attempts from the /student_exercise_history endpoint.

**Comment by SaraTolstrupHusum** [comment]
It seems to work just fine.
